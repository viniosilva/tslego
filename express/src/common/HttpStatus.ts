export enum HttpStatus {
  'OK' = 200,
  'CREATED' = 201,
  'NO_CONTENT' = 204,
  'NOT_FOUND' = 404,
  'INTERNAL_SERVER_ERROR' = 500,
}
