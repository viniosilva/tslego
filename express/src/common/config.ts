import { version, description } from '../../package.json';
import { config } from 'dotenv';
config();

export const apiConfig = {
  host: String(process.env.API_HOST),
  port: Number(process.env.API_PORT),
  swaggerConfig: {
    swaggerDefinition: {
      info: { title: 'TSLego Express Hello', version, description },
      host: `${String(process.env.API_HOST)}:${Number(process.env.API_PORT)}`,
      basePath: '/api',
    },
    apis: ['./src/user/**/*.ts'],
  },
};
