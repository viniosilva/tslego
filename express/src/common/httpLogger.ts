import { ServerResponse } from 'http';
import pinoHttp from 'pino-http';
import { logger } from './logger';

export function customLogLevel(res: ServerResponse, err: Error) {
  if (res.statusCode >= 400 && res.statusCode < 500) return 'warn';
  else if (res.statusCode >= 500 || err) return 'error';
  return 'info';
}

export const httpLogger = pinoHttp({
  logger,
  customLogLevel,
});

// Ref: https://www.npmjs.com/package/pino-http
