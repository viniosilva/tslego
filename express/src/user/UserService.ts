import Joi from 'joi';
import { User, UserInput } from './dto/User';
import { NotFoundException } from './exception/NotFoundException';

export class UserService {
  private readonly users: User[] = [];

  getUsers(): User[] {
    return this.users;
  }

  getUserById(userId: number): User {
    const user = this.users.find(({ id }) => id === userId);
    if (!user) throw new NotFoundException('User not found');
    return user;
  }

  createUser(userInput: UserInput): User {
    const user = new User();
    user.id = this.users.length + 1;
    user.name = userInput.name;

    this.users.push(user);
    return user;
  }

  updateUserById(userId: number, userInput: UserInput): User {
    const user = this.getUserById(userId);
    user.name = userInput.name;
    return user;
  }

  removeUserById(userId: number): void {
    let indexOf: number;
    this.users.some((user, index) => {
      if (user.id === userId) {
        indexOf = index;
        return true;
      }
    });

    if (!isNaN(indexOf)) this.users.splice(indexOf, 1);
  }

  validateUserInput(userInput: UserInput): void {
    const schema = Joi.object({
      name: Joi.string().min(3).max(100).required(),
    });
    const { error } = schema.validate(userInput);
    if (error)
      throw new Error(error.details.map(({ message }) => message).join(', '));
  }
}
