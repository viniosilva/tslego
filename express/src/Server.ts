import express, {
  NextFunction,
  Request,
  RequestHandler,
  Response,
} from 'express';
import helmet from 'helmet';
import { Logger } from 'pino';
import swaggerJSDoc from 'swagger-jsdoc';
import swaggerUi, { SwaggerOptions } from 'swagger-ui-express';
import { httpLogger } from './common/httpLogger';
import { HttpStatus } from './common/HttpStatus';

export class Server {
  readonly app = express();
  constructor(
    private readonly host: string,
    private readonly port: number,
    private readonly logger: Logger,
    swaggerConfig: SwaggerOptions,
    routes: RequestHandler[],
  ) {
    this.app.use(helmet());
    this.app.use(express.json());
    this.app.use(httpLogger);
    this.configureSwagger(swaggerConfig);

    routes.forEach((route) => this.app.use('/api', route));

    this.app.use(this.errorHandler.bind(this));
  }

  private configureSwagger(config: SwaggerOptions) {
    if (JSON.stringify(config) !== '{}') {
      this.app.use(
        '/api/docs',
        swaggerUi.serve,
        swaggerUi.setup(swaggerJSDoc(config)),
      );
    }
  }

  private errorHandler(
    err: Error,
    _req: Request,
    res: Response,
    _next: NextFunction,
  ) {
    this.logger.child({ errorStack: err.stack }).error(err.message);
    res.status(HttpStatus.INTERNAL_SERVER_ERROR).send('Internal server error!');
  }

  start() {
    this.app.listen(this.port, () => {
      this.logger.info(`App listening at [${this.host}:${this.port}]`);
    });
  }
}
