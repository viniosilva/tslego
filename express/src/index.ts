import { apiConfig } from './common/config';
import { logger } from './common/logger';
import { Server } from './server';
import { userController } from './user';

const routes = [userController.router];
const server = new Server(
  apiConfig.host,
  apiConfig.port,
  logger,
  apiConfig.swaggerConfig,
  routes,
);

server.start();
