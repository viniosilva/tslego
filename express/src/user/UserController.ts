import { NextFunction, Request, Response, Router } from 'express';
import { HttpStatus } from '../common/HttpStatus';
import { NotFoundException } from './exception/NotFoundException';
import { UserService } from './UserService';

/**
 * @swagger
 * tags:
 *   - name: User
 *     description: Operations about users
 */
export class UserController {
  readonly router = Router();

  constructor(private readonly userService: UserService) {
    this.router.get('/users', this.getUsers.bind(this));
    this.router.get('/users/:userId', this.getUserById.bind(this));
    this.router.post(
      '/users',
      this.validateUserInput.bind(this),
      this.createUser.bind(this),
    );
    this.router.put(
      '/users/:userId',
      this.validateUserInput.bind(this),
      this.updateUserById.bind(this),
    );
    this.router.delete('/users/:userId', this.removeUserById.bind(this));
  }

  /**
   * @swagger
   * /users:
   *   get:
   *     tags: ["User"]
   *     description: Returns the users
   *     responses:
   *       "200":
   *         description: "Successful operation"
   *         schema:
   *           type: "array"
   *           items:
   *             $ref: "#/definitions/User"
   */
  private getUsers(_req: Request, res: Response): void {
    const users = this.userService.getUsers();
    res.json(users);
  }

  /**
   * @swagger
   * /users/{userId}:
   *   get:
   *     tags: ["User"]
   *     description: Returns found user by id
   *     parameters:
   *     - name: "userId"
   *       in: "path"
   *       description: "ID of user that needs to be fethed"
   *       required: true
   *       type: "integer"
   *     responses:
   *       "200":
   *         description: "Successful operation"
   *         schema:
   *           $ref: "#/definitions/User"
   *       "404":
   *         description: "Not found"
   */
  private getUserById(req: Request, res: Response): void {
    try {
      const { userId } = req.params;
      const user = this.userService.getUserById(Number(userId));

      res.json(user);
    } catch (err) {
      if (err instanceof NotFoundException) {
        res.status(HttpStatus.NOT_FOUND).send(err.message);
      } else throw err;
    }
  }

  /**
   * @swagger
   * /users:
   *   post:
   *     tags: ["User"]
   *     description: Create user
   *     parameters:
   *     - name: "body"
   *       in: "body"
   *       description: "Create user"
   *       required: true
   *       schema:
   *         $ref: "#/definitions/User"
   *     responses:
   *       "201":
   *         description: "Successful operation"
   *         schema:
   *           $ref: "#/definitions/User"
   *       "400":
   *         description: "Bad request"
   */
  private createUser(req: Request, res: Response): void {
    const user = this.userService.createUser(req.body);
    res.status(HttpStatus.CREATED).json(user);
  }

  /**
   * @swagger
   * /users/{userId}:
   *   put:
   *     tags: ["User"]
   *     description: Update user by id
   *     parameters:
   *     - name: "userId"
   *       in: "path"
   *       description: "ID of user that needs to be update"
   *       required: true
   *       type: "integer"
   *     - name: "body"
   *       in: "body"
   *       description: "Update user"
   *       required: true
   *       schema:
   *         $ref: "#/definitions/User"
   *     responses:
   *       "200":
   *         description: "Successful operation"
   *         schema:
   *           $ref: "#/definitions/User"
   *       "400":
   *         description: "Bad request"
   *       "404":
   *         description: "Not found"
   */
  private updateUserById(req: Request, res: Response): void {
    try {
      const { userId } = req.params;
      const user = this.userService.updateUserById(Number(userId), req.body);

      res.json(user);
    } catch (err) {
      if (err instanceof NotFoundException) {
        res.status(HttpStatus.NOT_FOUND).send(err.message);
      } else throw err;
    }
  }

  /**
   * @swagger
   * /users/{userId}:
   *   delete:
   *     tags: ["User"]
   *     description: `Remove found user by id`
   *     parameters:
   *     - name: "userId"
   *       in: "path"
   *       description: "ID of user that needs to be update"
   *       required: true
   *       type: "integer"
   *     responses:
   *       "200":
   *         description: "Successful operation"
   */
  private removeUserById(req: Request, res: Response): void {
    const { userId } = req.params;
    this.userService.removeUserById(Number(userId));

    res.sendStatus(HttpStatus.NO_CONTENT);
  }

  private validateUserInput(req: Request, res: Response, next: NextFunction) {
    try {
      this.userService.validateUserInput(req.body);
      next();
    } catch (err) {
      res.status(HttpStatus.BAD_REQUEST).json({ error: err.message });
    }
  }
}
