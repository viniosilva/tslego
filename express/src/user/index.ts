import { UserController } from './UserController';
import { UserService } from './UserService';

const userService = new UserService();
export const userController = new UserController(userService);
