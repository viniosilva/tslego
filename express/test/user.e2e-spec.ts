import request, { Response } from 'supertest';
import { logger } from '../src/common/logger';
import { Server } from '../src/Server';
import { userController } from '../src/user';

describe('User E2E', () => {
  const server = new Server('', 0, logger, {}, [userController.router]);
  const path = '/api/users';

  describe('GET /api/users', () => {
    it('should return empty user list', () => {
      return request(server.app)
        .get(path)
        .expect(200)
        .expect((res: Response) => {
          expect(res.text).toMatchSnapshot();
        });
    });

    it('should return user list', async () => {
      const userInput = { name: 'Test name' };
      await request(server.app).post(path).send(userInput);

      return request(server.app)
        .get(path)
        .expect(200)
        .expect((res: Response) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('GET /api/users/:userId', () => {
    it('should return user', () => {
      return request(server.app)
        .get(`${path}/1`)
        .expect(200)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should throw not found exception', () => {
      return request(server.app)
        .get(`${path}/0`)
        .expect(404)
        .expect((res: Response) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('POST /api/users', () => {
    it('should return new user', () => {
      const userInput = { name: 'Test name' };
      return request(server.app)
        .post(path)
        .send(userInput)
        .expect(201)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });
  });

  describe('PUT /api/users/:userId', () => {
    it('should return updated user', () => {
      const userInput = { name: 'Test update name' };
      return request(server.app)
        .put(`${path}/1`)
        .send(userInput)
        .expect(200)
        .expect((res: Response) => {
          expect(res.body).toMatchSnapshot();
        });
    });

    it('should throw not found exception', () => {
      const userInput = { name: 'Test update name' };
      return request(server.app)
        .put(`${path}/0`)
        .send(userInput)
        .expect(404)
        .expect((res: Response) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('DELETE /api/users/:userId', () => {
    it('should remove user', () => {
      return request(server.app)
        .delete(`${path}/1`)
        .expect(204)
    });
  });
});
