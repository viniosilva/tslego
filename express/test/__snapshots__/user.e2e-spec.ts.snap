// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User E2E GET /api/users should return empty user list 1`] = `"[]"`;

exports[`User E2E GET /api/users should return user list 1`] = `"[{\\"id\\":1,\\"name\\":\\"Test name\\"}]"`;

exports[`User E2E GET /api/users/:userId should return user 1`] = `
Object {
  "id": 1,
  "name": "Test name",
}
`;

exports[`User E2E GET /api/users/:userId should throw not found exception 1`] = `"User not found"`;

exports[`User E2E POST /api/users should return new user 1`] = `
Object {
  "id": 2,
  "name": "Test name",
}
`;

exports[`User E2E POST /api/users should throw bad request exception 1`] = `
Object {
  "error": "\\"name\\" must be a string",
}
`;

exports[`User E2E PUT /api/users/:userId should return updated user 1`] = `
Object {
  "id": 1,
  "name": "Test update name",
}
`;

exports[`User E2E PUT /api/users/:userId should throw bad request exception 1`] = `
Object {
  "error": "\\"name\\" must be a string",
}
`;

exports[`User E2E PUT /api/users/:userId should throw not found exception 1`] = `"User not found"`;
